### PR TITLE
forward support for other types of devices (non-PCI) plus relevant cleanup

### LINKED ISSUE #13: allow device types other than PCI

### PR DESCRIPTION
**Pull request by tmakatos**
closes #13, #17

### CODE REVIEW

File: lib/libmuser.c [hunk 1]
```diff
@@ -77,16 +77,21 @@ typedef struct {
     _Static_assert(sizeof(s) - offsetof(s, m) == sizeof(t), \
         #t " " #m " must be last member in " #s)
 
+struct pci {
```
Comment (franciozzy) [inline lib/libmuser.c]: Should this be `lm_ctx_pci`?

File: lib/libmuser.c [hunk 2]
```diff
@@ -77,16 +77,21 @@ typedef struct {
...
     void                    *pvt;
     dma_controller_t        *dma;
     int                     fd;
     int (*reset)            (void *pvt);
     lm_log_lvl_t            log_lvl;
     lm_log_fn_t             *log;
-    lm_pci_info_t           pci_info;
-    lm_pci_config_space_t   *pci_config_space;
-    struct caps             *caps;
+    struct pci              pci;
```
Comment (franciozzy) [inline lib/libmuser.c]: Do we want to put this in a union already, or leave it for later?

File: lib/libmuser.c [hunk 3]
```diff
@@ -1207,8 +1212,13 @@ lm_ctx_create(const lm_dev_info_t *dev_info)
     lm_ctx->log_lvl = dev_info->log_lvl;
     lm_ctx->reset = dev_info->reset;
 
-    // Bounce the provided pci_info into the context.
-    memcpy(&lm_ctx->pci_info, &dev_info->pci_info, sizeof(lm_pci_info_t));
+    // Bounce info for IRQs and regions.
+    assert(sizeof(lm_ctx->pci.irq_count) == sizeof(dev_info->pci_info.irq_count));
```
Comment (franciozzy) [inline lib/libmuser.c]: These should be static asserts closer to the definition of these structs. You only want the test to happen during compilation, not at runtime every time.

File: lib/libmuser.c [hunk 4]
```diff
@@ -1207,8 +1212,13 @@ lm_ctx_create(const lm_dev_info_t *dev_info)
...
+    memcpy(lm_ctx->pci.irq_count, dev_info->pci_info.irq_count,
+           sizeof(lm_ctx->pci.irq_count));
+    assert(sizeof(lm_ctx->pci.reg_info) == sizeof(dev_info->pci_info.reg_info));
```
Comment (franciozzy) [inline lib/libmuser.c]: Same here.

File: lib/muser.h [hunk 5]
```diff
@@ -165,6 +165,45 @@ enum {
...
+     */
+    size_t size;
+
+    /*
+     * Function to call back when the capability gets read or written.
+     */
+    lm_cap_access_t *fn;
+} lm_cap_t;
+
+#define LM_MAX_CAPS (PCI_CFG_SPACE_SIZE - PCI_STD_HEADER_SIZEOF) / PCI_CAP_SIZEOF
```
Comment (franciozzy) [inline lib/muser.h]: Math in defines (the whole thing) should always be within brackets.
(I can see this was already wrong, but this is a good opportunity to fix it.)

File: lib/muser.h [hunk 6]
```diff
@@ -188,6 +227,19 @@ typedef struct {
      * Class code.
      */
     lm_pci_hdr_cc_t     cc;
+
+    /*
+     * PCI capabilities. The user needs to only define the ID and size of each
+     * capability. The actual capability is not maintained by libmuser. When a
+     * capability is accessed the appropriate callback function is called.
+     */
+    lm_cap_t        caps[LM_MAX_CAPS];
```
Comment (franciozzy) [inline lib/muser.h]: I think the alignment got lost, or there's a mix of spaces and tabs (not sure how github displays it).

For struct definitions where there's a lot of comments between elements, I'm not sure if there's value in trying to align everything. Maybe we want to "copy" (adapt) our internal "C" CODING_STYLE to this project.

File: lib/libmuser.c (hunk 1 above)
Comment (tmakatos) [reply lib/libmuser.c]: This struct is internal to the library, so there's no problem of namespace clash. Do we want to prefix internal structs as well?

File: lib/libmuser.c (hunk 2 above)
Comment (tmakatos) [reply lib/libmuser.c]: Let's leave it for later, it makes no difference.

File: lib/libmuser.c (hunk 3 above)
Comment (tmakatos) [reply lib/libmuser.c]: I wanted to do it as you suggest however I don't know how to statically assert that a member of a struct is of a specific size.

File: lib/libmuser.c (hunk 4 above)
Comment (tmakatos) [reply lib/libmuser.c]: Ditto.

File: lib/muser.h (hunk 6 above)
Comment (tmakatos) [reply lib/muser.h]: OK will remove alignment for those.

**Comment by tmakatos** [comment]
@franciozzy I've addressed your comments, please take another look.

**Comment by tmakatos** [comment]
Superseded by https://github.com/nutanix/libvfio-user/issues/106.

**Pull request closed without merging**